File: activitypub/person.go
package activitypub

import (
	"fmt"
	"time"

	"github.com/lacolaco/activitypub.lacolaco.net/model"
	"github.com/lacolaco/activitypub.lacolaco.net/sign"
	"humungus.tedunangst.com/r/webs/httpsig"
)

func NewPersonJSON(u *model.User, baseUri string, publicKey *httpsig.PublicKey) map[string]interface{} {
	apID := u.GetActivityPubID(baseUri)
	publicKeyPem, err := httpsig.EncodeKey(publicKey.Key)
	if err != nil {
		panic(err)
	}

	return map[string]interface{}{
		"@context":                  contextURIs,
		"type":                      "Person",
		"id":                        apID,
		"name":                      u.Name,
		"preferredUsername":         u.PrefName,
		"summary":                   u.Description,
		"inbox":                     fmt.Sprintf("%s/inbox", apID),
		"outbox":                    fmt.Sprintf("%s/outbox", apID),
		"followers":                 fmt.Sprintf("%s/followers", apID),
		"following":                 fmt.Sprintf("%s/following", apID),
		"url":                       fmt.Sprintf("%s/@%s", baseUri, u.ID),
		"published":                 u.CreatedAt.Format(time.RFC3339),
		"updated":                   u.UpdatedAt.Format(time.RFC3339),
		"discoverable":              true,
		"manuallyApprovesFollowers": false,
		"icon": map[string]interface{}{
			"type":      "Image",
			"mediaType": u.Icon.MediaType,
			"url":       u.Icon.URL,
		},
		"publicKey": map[string]interface{}{
			"id":           fmt.Sprintf("%s#%s", apID, sign.DefaultPublicKeyID),
			"owner":        apID,
			"publicKeyPem": publicKeyPem,
		},
	}
}

File: packages/server/src/activitypub/person.ts
import { User } from '@app/domain/user';
import { contextURIsWithExtensions } from './context';
import { buildPropertyValue } from './property-value';
import { Person, URI } from './schema';
import { getPublicKeyID } from './signature';

export function buildPerson(origin: string, user: User, publicKey?: string) {
  // `preferredUsername` はあとから変更可能にするため、不変なURIを `id` として使う
  const userURI = `${origin}/users/${user.id}`;

  return Person.parse({
    '@context': contextURIsWithExtensions,
    id: userURI,
    type: 'Person',
    name: user.displayName,
    preferredUsername: user.username,
    summary: user.description ?? '',
    icon: {
      type: 'Image',
      url: user.icon.url,
    },
    inbox: `${userURI}/inbox`,
    outbox: `${userURI}/outbox`,
    followers: `${userURI}/followers`,
    following: `${userURI}/following`,
    endpoints: {
      sharedInbox: new URL(`${origin}/inbox`),
    },
    url: user.url,
    published: user.createdAt,
    updated: user.updatedAt,
    attachment: user.attachments?.map((a) => buildPropertyValue(a.name, a.value)),
    manuallyApprovesFollowers: false,
    discoverable: true,
    publicKey: publicKey && {
      id: getPublicKeyID(userURI),
      owner: userURI,
      publicKeyPem: publicKey,
    },
  });
}

export async function fetchPersonByID(id: URI): Promise<Person> {
  const res = await fetch(id, {
    method: 'GET',
    headers: { Accept: 'application/activity+json, application/json' },
  });
  if (res.status !== 200) {
    throw new Error(`unexpected status code: ${res.status}`);
  }
  const body = await res.json();
  const parsed = Person.safeParse(body);
  if (parsed.success === false) {
    console.error(parsed.error);
    throw new Error(`unexpected object`);
  }
  return parsed.data;
}
